v2/pkg/utils/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,62 @@
+package ratelimit
+
+import (
+	"math"
+	"time"
+)
+
+// Limiter allows a burst of request during the defined duration
+type Limiter struct {
+	maxCount int
+	count    int
+	ticker   *time.Ticker
+	tokens   chan struct{}
+}
+
+func (limiter *Limiter) run() {
+	for {
+		if limiter.count <= 0 {
+			<-limiter.ticker.C
+			limiter.count = limiter.maxCount
+		}
+
+		select {
+		case limiter.tokens <- struct{}{}:
+			limiter.count--
+
+		case <-limiter.ticker.C:
+			limiter.count = limiter.maxCount
+		}
+	}
+}
+
+// Take one token from the bucket
+func (rateLimiter *Limiter) Take() {
+	<-rateLimiter.tokens
+}
+
+// New creates a new limiter instance with the tokens amount and the interval
+func New(max int, duration time.Duration) *Limiter {
+	limiter := &Limiter{
+		maxCount: max,
+		count:    max,
+		ticker:   time.NewTicker(duration),
+		tokens:   make(chan struct{}),
+	}
+	go limiter.run()
```

Comment:
this goroutine is started but there's no way to close the limited. We should create a cancellable context and have a stop function to stop the limiter releasing created resources

Reply:
Added support for external context, so the caller can use cancellable contexts (deadlines, timeout, etc)